### ISSUE TITLE
Error: Invalid Input Error: IllegalArgumentException: point array must contain 0 or >1 elements

### ISSUE THREAD

**Issue by marklit** [duckdblabs/duckdb_spatial#75]
```bash
$ wget https://usaminedroads.blob.core.windows.net/road-detections/USA.zip
$ unzip -j USA.zip
$ mv _USA.tsv USA.geojson
$ ogr2ogr -f GPKG USA.gpkg USA.geojson
```

```bash
$ wget -O tl_2012_us_state.zip \
    'https://www.sciencebase.gov/catalog/file/get/52c78623e4b060b9ebca5be5?facet=tl_2012_us_state'
$ unzip tl_2012_us_state.zip
```

```bash
$ wget https://biogeo.ucdavis.edu/data/diva/adm/USA_adm.zip
$ unzip USA_adm.zip
```

```bash
$ ~/duckdb_spatial/build/release/duckdb \
    -unsigned \
    USA.duckdb
```

```sql
INSTALL parquet;
LOAD parquet;
LOAD h3;

CREATE OR REPLACE TABLE states AS
    SELECT * EXCLUDE (wkb_geometry),
           ST_GEOMFROMWKB(wkb_geometry) AS geom
    FROM ST_READ('tl_2012_us_state.shp');

CREATE OR REPLACE TABLE counties AS
    SELECT NAME_2 AS county_name,
           ST_GEOMFROMWKB(wkb_geometry) AS geom
    FROM ST_READ('USA_adm2.shp');

COPY (SELECT ST_AsHEXWKB(a.geom::GEOMETRY)::TEXT AS geom,
             b.OBJECTID AS OBJECTID,
             b.REGION AS REGION,
             b.DIVISION AS DIVISION,
             b.STATEFP AS STATEFP,
             b.STATENS AS STATENS,
             b.GEOID AS GEOID,
             b.STUSPS AS STUSPS,
             b.NAME AS NAME,
             b.LSAD AS LSAD,
             b.MTFCC AS MTFCC,
             b.FUNCSTAT AS FUNCSTAT,
             b.ALAND AS ALAND,
             b.AWATER AS AWATER,
             b.INTPTLAT AS INTPTLAT,
             b.INTPTLON AS INTPTLON,
             b.Shape_Leng AS Shape_Leng,
             b.Shape_Area AS Shape_Area,
             c.county_name AS county_name,
             printf('%x', h3_latlng_to_cell(
                               ST_Y(ST_CENTROID(ST_GEOMFROMWKB(a.geom))),
                               ST_X(ST_CENTROID(ST_GEOMFROMWKB(a.geom))),
                               7)::bigint) as h3_7,
             printf('%x', h3_latlng_to_cell(
                               ST_Y(ST_CENTROID(ST_GEOMFROMWKB(a.geom))),
                               ST_X(ST_CENTROID(ST_GEOMFROMWKB(a.geom))),
                               8)::bigint) as h3_8,
             printf('%x', h3_latlng_to_cell(
                               ST_Y(ST_CENTROID(ST_GEOMFROMWKB(a.geom))),
                               ST_X(ST_CENTROID(ST_GEOMFROMWKB(a.geom))),
                               9)::bigint) as h3_9
      FROM st_read('USA.gpkg') a
      LEFT JOIN states   b ON ST_WITHIN(ST_GEOMFROMWKB(a.geom), b.geom)
      LEFT JOIN counties c ON ST_WITHIN(ST_GEOMFROMWKB(a.geom), c.geom))
TO 'USA.pq' (FORMAT 'PARQUET',
             CODEC  'Snappy');
```

```
Error: Invalid Input Error: IllegalArgumentException: point array must contain 0 or >1 elements
```

Any idea what the above error might be referring to? I can make the above JOIN work on a single record so I suspect there is some bad geometry in the `USA.gpkg` file. Any idea how to isolate it?

```sql
D create table test as select * FROM st_read('USA.gpkg') a limit 1;
D select * from test;
```

```
┌─────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────┐
│                                                          geom                                                           │
│                                                        wkb_blob                                                         │
├─────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────┤
│ LINESTRING (-105.528080463409 48.8574728849071, -105.527983903885 48.857451708292, -105.526428222656 48.8564211021993…  │
└─────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────┘
```

```sql
SELECT ST_AsHEXWKB(a.geom::GEOMETRY)::TEXT AS geom,
       b.OBJECTID AS OBJECTID,
       b.REGION AS REGION,
       b.DIVISION AS DIVISION,
       b.STATEFP AS STATEFP,
       b.STATENS AS STATENS,
       b.GEOID AS GEOID,
       b.STUSPS AS STUSPS,
       b.NAME AS NAME,
       b.LSAD AS LSAD,
       b.MTFCC AS MTFCC,
       b.FUNCSTAT AS FUNCSTAT,
       b.ALAND AS ALAND,
       b.AWATER AS AWATER,
       b.INTPTLAT AS INTPTLAT,
       b.INTPTLON AS INTPTLON,
       b.Shape_Leng AS Shape_Leng,
       b.Shape_Area AS Shape_Area,
       c.county_name AS county_name,
       printf('%x', h3_latlng_to_cell(
                         ST_Y(ST_CENTROID(ST_GEOMFROMWKB(a.geom))),
                         ST_X(ST_CENTROID(ST_GEOMFROMWKB(a.geom))),
                         7)::bigint) as h3_7,
       printf('%x', h3_latlng_to_cell(
                         ST_Y(ST_CENTROID(ST_GEOMFROMWKB(a.geom))),
                         ST_X(ST_CENTROID(ST_GEOMFROMWKB(a.geom))),
                         8)::bigint) as h3_8,
       printf('%x', h3_latlng_to_cell(
                         ST_Y(ST_CENTROID(ST_GEOMFROMWKB(a.geom))),
                         ST_X(ST_CENTROID(ST_GEOMFROMWKB(a.geom))),
                         9)::bigint) as h3_9
FROM test a 
LEFT JOIN states   b ON ST_WITHIN(ST_GEOMFROMWKB(a.geom), b.geom)
LEFT JOIN counties c ON ST_WITHIN(ST_GEOMFROMWKB(a.geom), c.geom);
```

```
┌──────────────────────┬──────────┬─────────┬──────────┬───┬─────────────────┬─────────────────┬─────────────────┐
│         geom         │ OBJECTID │ REGION  │ DIVISION │ … │      h3_7       │      h3_8       │      h3_9       │
│       varchar        │  int32   │ varchar │ varchar  │   │     varchar     │     varchar     │     varchar     │
├──────────────────────┼──────────┼─────────┼──────────┼───┼─────────────────┼─────────────────┼─────────────────┤
│ 010200000007000000…  │          │         │          │ … │ 872785d13ffffff │ 882785d13dfffff │ 892785d13d3ffff │
├──────────────────────┴──────────┴─────────┴──────────┴───┴─────────────────┴─────────────────┴─────────────────┤
│ 1 rows                                                                                    22 columns (7 shown) │
└────────────────────────────────────────────────────────────────────────────────────────────────────────────────┘
```

`ST_ISVALID` brings the issue up as well.

```sql
D select st_isvalid(geom::GEOMETRY), count(*) from st_read('USA.gpkg') group by 1;
```

```
Error: Invalid Input Error: IllegalArgumentException: point array must contain 0 or >1 elements
```

I also tried to see if there was some very short geometry that might be triggering this but the following attempt failed.

```sql
D select geom from st_read('USA.gpkg') order by length(geom::TEXT) limit 50;
```

```
Error: Invalid Input Error: IllegalArgumentException: point array must contain 0 or >1 elements
```

**Comment by marklit** [comment]
I put together a more self-contained example:

```sql
select * from 'train.csv' LIMIT 1;
```

```
     TRIP_ID = 1372636858620000589
   CALL_TYPE = C
 ORIGIN_CALL = 
ORIGIN_STAND = 
     TAXI_ID = 20000589
   TIMESTAMP = 1372636858
    DAY_TYPE = A
MISSING_DATA = false
    POLYLINE = [[-8.618643,41.141412],[-8.618499,41.141376],[-8.620326,41.14251],[-8.622153,41.143815],[-8.623953,41.144373],[-8.62668,41.144778],[-8.627373,41.144697],[-8.630226,41.14521],[-8.632746,41.14692],[-8.631738,41.148225],[-8.629938,41.150385],[-8.62911,41.151213],[-8.629128,41.15124],[-8.628786,41.152203],[-8.628687,41.152374],[-8.628759,41.152518],[-8.630838,41.15268],[-8.632323,41.153022],[-8.631144,41.154489],[-8.630829,41.154507],[-8.630829,41.154516],[-8.630829,41.154498],[-8.630838,41.154489]]
```

```sql
SELECT CONCAT('LINESTRING(', replace(replace(regexp_replace(POLYLINE, '[0-9],[0-9]', '\1 \2', 'g'), '[', ''), ']', ''), ')') geom from 'train.csv' LIMIT 1;
```

```
 geom = LINESTRING(-8.61864 1.141412,-8.61849 1.141376,-8.62032 1.14251,-8.62215 1.143815,-8.62395 1.144373,-8.6266 1.144778,-8.62737 1.144697,-8.63022 1.14521,-8.63274 1.14692,-8.63173 1.148225,-8.62993 1.150385,-8.6291 1.151213,-8.62912 1.15124,-8.62878 1.152203,-8.62868 1.152374,-8.62875 1.152518,-8.63083 1.15268,-8.63232 1.153022,-8.63114 1.154489,-8.63082 1.154507,-8.63082 1.154516,-8.63082 1.154498,-8.63083 1.154489)
```

```sql
SELECT ST_GEOMFROMTEXT(CONCAT('LINESTRING(', replace(replace(regexp_replace(POLYLINE, '[0-9],[0-9]', '\1 \2', 'g'), '[', ''), ']', ''), ')')) geom FROM 'train.csv' LIMIT 1;
```

```
Error: Invalid Input Error: IllegalArgumentException: point array must contain 0 or >1 elements
```

**Comment by marklit** [comment]
For the MS Roads dataset, there are [204,789 records](https://github.com/microsoft/RoadDetections/issues/11) with a single point in their `LINESTRING`. I'm not sure how to filter these out ahead of importing into DuckDB. I'm able to import them into PostGIS and then use `ST_NumPoints` when cleaning up the table.

**Comment by Maxxen** [comment]
This is another issue where invalid geometries (in this case linestrings with 1 point) gets fed to GEOS when converting to WKT (which we do when printing), which is unable to handle such geometries.

One way to circumvent this is by casting the read geometry into a `LINESTRING_2D` which uses another code path to print as text, or otherwise filter the invalid geometry before printing/feeding through GEOS. Although we would probably have to implement `ST_NumPoints` and/or check the number of points in `ST_IsValid` before calling into GEOS. I'll see what I can do. 

Using our own code for printing instead of GEOS is also high on my list.

**Comment by Maxxen** [comment]
Most of these issues should be solved in #85.

**Comment by marklit** [comment]
Thanks for adding those native functions.

The dataset I was seeing these issues on needed fairly extensive ETL to get cleaned up, I detailed what I did here: https://github.com/microsoft/RoadDetections/issues/11#issuecomment-1576013896

If/when #81 is implemented and there is some way to filter out unique points from LINESTRINGs in DuckDB I could see if the Shapely code I wrote could be ported to DuckDB entirely. As for now, I'm processing the file entirely in Python. It could be a DuckDB UDF could be a middle ground.
